### PR TITLE
linear_cross_entropy: support probability targets on the chunked path

### PR DESCRIPTION
**Pull request by pearu**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.13.0) (oldest at bottom):
* #187838
* __->__ #187053

Probability (soft-label) targets previously always fell back to the
reference path. This adds chunked support for reduction='mean'/'sum'
when the target dtype matches the input dtype and the target does not
require grad; other probability-target configurations (reduction='none',
target requiring grad, mismatched dtype) keep falling back with the
existing warning.

The scalar custom op is reused as-is: for scalar reductions the
autograd model (precompute grads in forward, scale by the scalar
upstream grad in backward) is unchanged, so no new op, registrations,
or xfail wiring -- the accumulator branches into a probability-target
loop. The loop is a dense variant of the index-target loop: the
per-row weighting becomes the (B, V) prob_wt (= class_weight * target),
the loss target term is a flat dot, the softmax scaling uses the
per-chunk row mass s_n = sum_c (w*t)[n, c], and the gradient
correction is a dense subtraction -- no gather / index_add_ /
index_select. The mean divisor is the row count, not the weight mass,
matching cross_entropy_loss_prob_target.

Testing reuses the _test_linear_cross_entropy_loss harness (a new
prob_target leg with per-device ULP caps -- fp32 is split by
device.type since its ULP spreads by host, while the fp16/bf16/accurate
caps are shared across devices), plus a prob_target axis on the chunked
gradcheck and a dispatch-edges test. The chunked probability-target
module samples are appended after the existing generator output so the
RNG draw order behind the calibrated index-target caps is unchanged.

fp16 robustness: two reductions in the probability-target loop return
fp16 scalars and could overflow fp16's 65504 ceiling -- the softmax
denominator (a sum of ~num_classes terms in (0, 1], overflowing for a
large vocabulary) and the loss target term sum_{n,c} (w*t)*x_shifted
(overflowing for a large batch of widely-spread early-training logits,
since cublasDotEx returns its scalar at CUDA_R_16F). Both now reduce in
fp32 for fp16 input, matching softmax / cross_entropy: the denominator
via sum(dtype=float32), the loss term via a per-row einsum summed across
rows in fp32 (measured fused -- same time and no extra peak vs
torch.dot). bf16 shares fp32's exponent range and fp32/fp64 already
suffice, so the widening is fp16-only. Regression tests cover both
overflow regimes.

Test Plan:

```
python test/test_nn.py -v -k "linear_cross_entropy and prob"
```

The fp32 ULP caps were calibrated per platform (CUDA measured directly;
ROCm, aarch64/x86_64 CPU, and MPS from CI), then set with margin.

This PR was authored with the assistance of an AI tool (Claude).

cc @albanD @mruberry @jbschlosser @walterddr @mikaylagawarecki

### CODE REVIEW

**Comment by pearu** [comment]
### A100 benchmark (bf16, probability targets): chunked auto-dispatch vs reference


auto_dispatch_bfloat16_NVIDIA_A100_SXM4_80GB_prob


`auto` (chunked) vs the full-materialization reference across `num_tokens` / `in_features` / `num_classes` (no liger -- its fused op is index-target only). Rows: time, peak memory, gradient rel error vs an fp64 reference.

**The accuracy row is the headline.** The chunked path keeps logits in fp32 accumulation; the reference rounds them to bf16 before the softmax. So chunked is **~3-4x more accurate** than the reference at every shape (grad rel-err ~2-3e-3 vs ~8-10e-3). That -- not only memory -- is why `auto` stays on the chunked path rather than the reference.

`num_classes` axis (`num_tokens=4096`, `in_features=4096`):

| num_classes | ref mem | auto mem | mem saving | ref grad-err | auto grad-err |
|---|---|---|---|---|---|
| 8192 | 0.42 | 0.45 | -- | 9.6e-3 | 2.4e-3 |
| 16384 | 0.80 | 0.63 | 22% | 9.7e-3 | 2.7e-3 |
| 32000 | 1.51 | 0.97 | 36% | 9.0e-3 | 2.3e-3 |
| 65536 | 3.05 | 1.74 | 43% | 7.7e-3 | 2.8e-3 |

Memory savings grow with `num_classes` -- the chunked logits buffer replaces the full `(num_tokens, num_classes)` materialization -- reaching ~43% at `num_classes=65536`; below `num_classes ~ in_features` chunking carries a small overhead instead. The cost is time: chunked runs ~1.3-1.7x slower (chunked matmuls plus the per-chunk softmax structure).

_Re-confirmed on the decoupled build (A100, head `fb396e43`): the numbers above reproduce within rounding (chunk-size pick unchanged after the rebase)._

**Comment by slayton58** [comment]
@claude Review this PR, exhaustively find all potential issues, concentrating on subtle problems a human reviewer might miss. Ignore small things, and concisely flag only major issues.

File: test/test_nn.py [hunk 1]
```diff
@@ -15140,6 +15174,25 @@ def grad_max_ulp(grad, expected):
                     maximal_linear_bias_grad_err = err
                     worst_linear_bias_grad_err_kwargs = dict(module_kwargs)
 
+        if prob_target:
+            # TEMP: calibration data for the probability-target ULP caps.
+            # Re-run per host, copy the observed values into the
+            # ``if prob_target:`` cap block above (keyed on policy /
+            # dtype / device), then delete this block.
```
Comment (slayton58) [inline test/test_nn.py]: Needs to be removed?

File: torch/nn/modules/linear_cross_entropy.py [hunk 2]
```diff
@@ -790,6 +876,70 @@ def _linear_cross_entropy_batch_chunked_accumulator(
         compute_input_grad or compute_linear_weight_grad or compute_linear_bias_grad
     )
 
+    if ctx.is_prob_target:
```
Comment (slayton58) [inline torch/nn/modules/linear_cross_entropy.py:886]: Factor out into a method (and perhaps the same for the non-prob-target loop below)?

File: torch/nn/modules/linear_cross_entropy.py [hunk 3]
```diff
@@ -790,6 +876,70 @@ def _linear_cross_entropy_batch_chunked_accumulator(
...
+        # Probability-target loop. The per-row weighting is the dense
```
Comment (slayton58) [inline torch/nn/modules/linear_cross_entropy.py:887]: Conceptually how different is this loop to the non-prob-target one? Can the code be unified?

**Review by slayton58 — COMMENTED** [review]
Generally looks good, general comments about code complexity / readability / DRY

File: torch/nn/modules/linear_cross_entropy.py (hunk 3 above)
Comment (pearu) [reply torch/nn/modules/linear_cross_entropy.py:887]: Both comments are addressed by the same change. I factored the shared forward prelude (`logits = input @ weight.T (+ bias)`, then the softmax-stability `-= row_max`) into a `_ChunkContext.shifted_logits(chunk)` method, now used by all three chunk loops (none / prob / index) -- so the duplication is gone, including the non-prob loop; the single-use `amax` helper it called is inlined there.

On unifying the full loop bodies, I kept them separate. The index loop is sparse (`dotgather` / `index_select` / `index_add_`, a one-hot target, negated `neg_weight_target` sign convention), while the prob loop is dense: a flat `(B, V)` dot, per-row mass `s_n`, and the gradient correction folded into `grad_logits` via `logits.sub_(wt)` before a clean matmul. Unifying would put an `if is_prob_target` branch at each of those ~4 points in the hot loop, which reads worse than two focused loops -- the differing parts are the substance.

File: torch/nn/modules/linear_cross_entropy.py (hunk 2 above)
Comment (pearu) [reply torch/nn/modules/linear_cross_entropy.py:886]: Addressed together with the unification question below: https://github.com/pytorch/pytorch/pull/187053#discussion_r3443518511

File: test/test_nn.py (hunk 1 above)
Comment (pearu) [reply test/test_nn.py]: yes, before the landing this if-block must be removed. Keeping it for now until ULP data correctness is confirmed.

**Comment by pearu** [comment]
`reduction='none'` with a probability target is a planned follow-up -- it falls through to the reference for now, and the chunked version needs the recompute-from-saved backward (a different approach from the `mean`/`sum` precomputed-grad path here, same as the existing index-target `none` case), so it's cleaner to add on its own.

File: test/test_nn.py (hunk 1 above)
Comment (pearu) [reply test/test_nn.py]: Done -- removed in the latest push, along with the rest of the calibration scaffolding (the `import sys`, the per-process counter, and its class attribute). The probability-target ULP caps are now finalized from cross-platform data: fp32 is split by `device.type` (cuda 512/256, cpu 16384/2048 driven by aarch64, mps 256/64), and the fp16/bf16/accurate caps cover every device. The cap-block comment documents the observed maxima per platform.

**Comment by slayton58** [comment]
@claude Re-review this PR, concentrating on subtle problems a human reviewer might miss. Ignore small things, and concisely flag only major issues.

**Comment by pearu** [comment]
Thanks — fixed. The fp16 prob loss-term `torch.dot(wt, logits)` overflowed because `cublasDotEx` returns its scalar at fp16; it now reduces per-row and then across rows in fp32 via `torch.einsum("bv,bv->b", wt, logits).sum(dtype=torch.float32)`, gated to the fp16 logits-buffer path (bf16/fp32 return a wide-enough scalar, so they keep the flat dot). I benchmarked the einsum form as fused — identical latency and **zero** extra peak memory vs `torch.dot` — so it costs the chunked memory bound nothing. Corrected the inline comment (the gain is overflow robustness, not accuracy).

While there I found and fixed the companion overflow in the softmax denominator (`sumexp_`): it sums ~`num_classes` terms in (0, 1], so it overflows fp16 for a large vocabulary — now floored to fp32 for fp16 input, matching what softmax / cross_entropy already do. Both are covered by new regression tests (`test_linear_cross_entropy_prob_large_magnitude_fp16_loss_term` and `…_prob_large_vocab_fp16_denom`), each verified to go `inf` without the fix.

File: torch/nn/modules/linear_cross_entropy.py [hunk 4]
```diff
@@ -591,21 +676,34 @@ def mm(self, mat1: torch.Tensor, mat2: torch.Tensor, *, out: torch.Tensor) -> No
         else:
             torch.mm(mat1, mat2, out_dtype=out.dtype, out=out)
 
-    def amax(self, x: torch.Tensor) -> torch.Tensor:
-        out = self.tmp.narrow(0, 0, x.shape[0]).unsqueeze(1)
-        torch.amax(x, dim=1, keepdim=True, out=out)
-        return out
+    def shifted_logits(self, chunk: "_ChunkViews") -> torch.Tensor:
```
Comment (slayton58) [inline torch/nn/modules/linear_cross_entropy.py]: Do we not have an actual definition of `_ChunkViews` here?

File: torch/nn/modules/linear_cross_entropy.py [hunk 5]
```diff
@@ -591,21 +676,34 @@ def mm(self, mat1: torch.Tensor, mat2: torch.Tensor, *, out: torch.Tensor) -> No
...
-        # on rows with widely-spread logits, after which log(0)=-inf
-        # poisons the loss and 1/0 the gradient.
-        return x.exp_().sum(dim, dtype=self.acc_dtype)
+        # Accumulate the normalizer in fp32 for fp16 input, like softmax /
+        # cross_entropy (which always reduce in acc_type): the denom sums
+        # num_classes terms in (0, 1] and overflows fp16 (>65504) for a large
+        # vocabulary. bf16 shares fp32's exponent range so it cannot overflow,
+        # and fp32/fp64 already suffice -- so the widening is fp16-only,
+        # keeping the fp32 radius minimal.
+        reduce_dtype = torch.float32 if self.dtype == torch.float16 else self.acc_dtype
```
Comment (slayton58) [inline torch/nn/modules/linear_cross_entropy.py:705]: Don't we care about extra mantissa bits for bf16 here? It's not just range, there's a precision concern too.

File: torch/nn/modules/linear_cross_entropy_options.py [hunk 6]
```diff
@@ -13,14 +13,19 @@ class _AutoDefault(NamedTuple):
...
-_AUTO_DEFAULTS: dict[tuple[str, torch.dtype], _AutoDefault] = {
-    ("cuda", torch.bfloat16): _AutoDefault("compact", "aspect_ratio:2"),
-    ("cuda", torch.float16): _AutoDefault("compact", "aspect_ratio:2"),
-    ("cpu", torch.bfloat16): _AutoDefault("accurate", "aspect_ratio"),
-    ("cpu", torch.float16): _AutoDefault("accurate", "aspect_ratio"),
+# keyed by ``(device_type, input_dtype, prob_target)`` -- Pareto picks from
+# an fp64-jacobian sweep. CPU picks ``"accurate"`` (only chunked policy with
+# fp32 weight-grad mm on CPU; others hit emulated low-precision matmul,
+# ~20-50x slower). Probability targets on CUDA pick aspect_ratio factor 1.
+_AUTO_DEFAULTS: dict[tuple[str, torch.dtype, bool], _AutoDefault] = {
```
Comment (slayton58) [inline torch/nn/modules/linear_cross_entropy_options.py:20]: `tuple` -> `NamedTuple` to make it more understandable throughout the code?

**Review by slayton58 — COMMENTED** [review]
Only the bf16 reduce_dtype actually causes me any concern, otherwise couple of nits, but we're basically there.

File: torch/nn/modules/linear_cross_entropy.py (hunk 5 above)
Comment (pearu) [reply torch/nn/modules/linear_cross_entropy.py:705]: fp16 accumulates the denom in fp32 (the agreed overflow fix); the question is whether bf16 should too, for precision. Measured: no benefit.

**Where this applies at all:** only when `acc_dtype` resolves to bf16 — an explicit non-`auto` `acc_policy` with bf16 (any device), or `auto`+bf16 on MPS / pre-SM80 CUDA. On `auto`+bf16 with SM80+ CUDA or CPU the denom is already accumulated in fp32, so it's moot there.

The denom `Σ_v exp(shifted_v)` (`shifted ≤ 0` ⇒ each term in (0, 1]) is consumed only as `log(denom)` and `1/denom`, both stored in bf16 on that path. So the test is whether computing it in fp32 vs bf16 changes those bf16-stored results:

```python
import torch
def ulp(a, b):  # bf16 ULPs between two values, once both are stored in bf16
    return (a.bfloat16().view(torch.int16).int() - b.bfloat16().view(torch.int16).int()).abs()
d  = torch.linspace(1., 131072., 4_000_000)   # true denom (fp32-accumulated)
db = d.bfloat16().float()                      # denom stored in bf16
print(f"log(denom): max {ulp(d.log(), db.log()).max()} bf16-ULP   "
      f"1/denom: max {ulp(1/d, 1/db).max()} bf16-ULP   "
      f"max |Δlog| = {(d.log()-db.log()).abs().max():.1e} (= ½·eps_bf16)")
```
```
log(denom): max 10 bf16-ULP   1/denom: max 1 bf16-ULP   max |Δlog| = 3.9e-3 (= ½·eps_bf16)
```

`1/denom` (the softmax/grad scale) is within **1 bf16 ULP** everywhere; `log(denom)` within 1 ULP except denom ∈ [1, 1.5), where the 10-ULP case is `log(denom) ≈ 0` (an almost one-hot softmax) — a ≤ 3.9e-3 absolute error. So computing the bf16 denom in fp32 moves the bf16-stored loss/softmax by **≤ 1 ULP** — nothing beyond bf16's own grid.

**Net:** switching bf16→fp32 buys nothing — no correctness gain (bf16 can't overflow), no measurable precision gain (≤1 ULP; random + adversarial samples show no net change), no perf/memory gain (the sum already accumulates in fp32 internally; only the `(B,)` denom's storage dtype changes). Keeping bf16 is positively preferable: it keeps the fp32 escalation tied to the one real failure mode (fp16 overflow) instead of special-casing without cause, avoids re-calibrating the bf16 ULP caps for a ≤1-ULP change, keeps `acc_dtype=bf16` actually meaning bf16, and is even marginally more accurate for flat-large-V (fp32-denom removes a cancelling bf16 error → ~4× worse loss there). So the fp32 accumulation stays fp16-only.

File: torch/nn/modules/linear_cross_entropy_options.py (hunk 6 above)
Comment (pearu) [reply torch/nn/modules/linear_cross_entropy_options.py:20]: This `_AUTO_DEFAULTS` dict -- its `tuple` key and the `_AutoDefault` NamedTuple -- is removed entirely in the stacked follow-up #187838, replaced by `_auto_acc_policy(device_type, dtype)` (a plain function, no lookup table or tuple key). So rather than NamedTuple-ify the key here only for #187838 to delete it, I'd suggest leaving it as-is and resolving this there. Happy to add the NamedTuple in this PR regardless if you'd prefer.

File: torch/nn/modules/linear_cross_entropy.py (hunk 4 above)
Comment (pearu) [reply torch/nn/modules/linear_cross_entropy.py]: Yes -- `_ChunkViews` is defined at L61, just above `_ChunkContext`, so it's already in scope here; the string was an unnecessary forward-ref. Dropped the quotes -> `chunk: _ChunkViews`. (`ctx: "_ChunkContext"` at L66 and the `-> "_ChunkContext"` return at L430 stay quoted -- those are genuine forward-refs, since `_ChunkContext` is defined later.)

**Review by slayton58 — APPROVED** [review]

File: torch/nn/modules/linear_cross_entropy_options.py (hunk 6 above)
Comment (slayton58) [reply torch/nn/modules/linear_cross_entropy_options.py:20]: I'm fine with the stacked removal - we'll merge this as a stack anyway
